### ISSUE TITLE
EventDispatcher - Listener Activation Check

### ISSUE THREAD

**Issue by webdevsHub** [symfony/symfony#14257]
I want to request a feature in EventDispatcher Component to speed up dispatching.

The typical Listener in a Symfony project looks like that:

```php
class SomeListener
{
    private $dep1;
    private $dep2;
    // ...
    private $depN;

    public function __construct($dep1, $dep2, /* ... */ $depN)
    {
        $this->dep1 = $dep1;
        $this->dep2 = $dep2;
        // ...
        $this->depN = $depN;
    }

    public function onEventName(SomeEvent $event)
    {
        if(!someStaticCondition($event)) {
            return;
        }

        // actual listener code
    }
}
```

So basically I want to execute `//actual listener code` when `EventName` is fired and `someStaticCondition` is true. In every other case Symfony should ignore the whole code at all.

But following actually happens:

1. Listener service definition is added via `Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher::addListenerService()`
2. `EventName` is fired and `ContainerAwareEventDispatcher` resolves every service listening to `EventName` and adds it to the underlying `Symfony\Component\EventDispatcher` base
3. `EventDispatcher` executes `SomeListener::onEventName()` via `Symfony\Component\EventDispatcher::doDispatch()`

I want to focus on the worst case: `EventName` is `kernel.request` and `someStaticCondition` is only true in 1% of all requests => The Listener + dependencies are instantiated by DI and not used in 99% of all requests because the condition is false.

The lazyLoad magic in `ContainerAwareEventDispatcher` can not increase performance in that example, because `kernel.request` is always fired, so every listener is fully resolved on every request.

Okay. To avoid unnecessary DI resolving I usually use Lazy Services. But that does not work here because `Symfony\Component\EventDispatcher::doDispatch()` always invokes the service. Why? Because `someStaticCondition` is mixed within the listener code.

My request is to standardize that wide spread `someStaticCondition` approach in a way that Listeners dependencies must not be resolved.

---- Possible solution ----

Listeners could implement an Interface with one static(!) condition like

```php
interface ListenerCondition
{
    /**
     * @return bool
     */
    public static function condition($event);
}
```

now `Symfony\Component\EventDispatcher::doDispatch()` could first check if a Listener implements that Interface and execute the static condition and continue if the condition fails.

If you use Lazy Services + the stuff I suggested developers can ensure that listeners are only touched (instantiated and executed) when `EventName` is fired AND `someStaticCondition` is true.

I can create a PR with that Interface, but I want to discuss my idea first.

**Comment by stof** [comment]
There is a big issue with your proposal:a class can contain as many listeners as you want, given that a listener is a callable, not an object.
So an interface with a static method being called does not work, because it is incompatible with classes containing multiple listeners (which happens regularly, even inside Symfony itself).
The only API I see to handle this would be to let the callable for the condition being registered in the event dispatcher alongside the listener callable. Introducing it in a BC way would be hard though. I don't see how it could be done without changing the EventDispatcherInterface.

Btw the ContainerAwareEventDispatcher cannot know the class of the listener to see whether it implements a given interface. All it has is the service id.

**Comment by webdevsHub** [comment]
That Interface implementation is only a suggestion therefore I did not created a PR.

The interface check would be done in `EventDispatcher::doDispatch()` not in `ContainerAwareEventDispatcher`, so no service ids.

For Subscribers I can think of an `SubscriberCondition` Interface or in general only support single class listeners with that condition (it would be optional anyway).

I like your idea to register a second condition-callable because you could add conditions even if your listener is a lambda. But as you mentioned it would break BC. For that reason I suggested the interface solution.

The main thing I want to do is to move the `someStaticCondition` check into a separate logic because it is used quite often and for me it looks very "hacky" to mix that check into the normal listener logic (). Any other Ideas to do that?

**Comment by jakzal** [comment]
@webdevsHub as a workaround you could implement the condition in a `kernel.request` listener, which would dispatch a custom event if the condition is met. Your main listener would listen to the custom event. 

Static condition methods would only solve the problem partially as there are cases when you'd like to use services to make a decision (without loading the unnecessary services).

I'm a bit afraid that introducing a built in mechanism for conditional listeners would complicate the way dispatcher works (it's already hard to follow for some).

**Comment by stof** [comment]
I'm totally against adding a feature which works only for single-listener classes (and being totally broken for other cases, without any sane way for us to warn the user). So the interface-based API is a no-go IMO.

**Comment by Rvanlaak** [comment]
Can't the multiple listener per class problem be solved by making the condition an optional parameter in the listener tag declaration? That way the Listener doesn't has to implement the interface.

```yml
app.demo_listener:
    class: AppBundle\EventListener\DemoListener
    arguments:
        - @translator
        ...
        - @security.http_utils
    tags:
        - { name: kernel.event_listener, event: security.interactive_login }
        - { name: kernel.event_listener, event: kernel.response, method: onKernelResponse }
        ...
        - { name: kernel.event_listener, event: kernel.response, method: onKernelAdminResponse, condition: staticAdminConditionMethod }
```

**Comment by dupuchba** [comment]
@Rvanlaak It is quite a good idea IMO.

**Comment by SoboLAN** [comment]
@Rvanlaak I like your idea too. Although I would be a bit cautios with this because encouraging developers to rely on static context is less than ideal (it discourages the use of DI, which in turn makes it harder to write unit-tests ) .

Maybe having something like this implemented but with a big red warning/recommendation in the documentation that it should only be used in the case when instantiating the listener is an expensive operation.

**Comment by Rvanlaak** [comment]
What about the condition not being static, it also could be a callable or other service definition right?

**Comment by SoboLAN** [comment]
Yes, but if you'd have to instantiate another service to test the condition, that would defeat the entire purpose of the feature (which is improving performance).

**Comment by jakzal** [comment]
@SoboLAN not necessarily. You might need a different set of services (more lightweight) to make a decision, than those needed for the actual listener. However, this can already be solved with an approach I gave earlier.

**Comment by Rvanlaak** [comment]
Introducing a custom event where the service would listen to as @jakzal proposes already is a neat solution, but adds more complexity to DI in my opinion. That way a custom event always is needed, while adding a condition could be anything between a simple static callable and a lightweight service. Especially for services like APIs that have network overhead this could improve performance.

**Comment by jakzal** [comment]
@Rvanlaak I'm not proposing to add anything to the framework. I only describe what you can already do.

**Comment by Rvanlaak** [comment]
@jakzal did understood that :+1:
